### ISSUE TITLE
TiDB deployment doesn't work on MacOS

### ISSUE THREAD

**Issue by datacharmer** [datacharmer/dbdeployer#62]
**Describe the bug**

I deploy TiDB using `dbdeployer deploy single tidb3.0.0 --client-from=5.7.25` 
If I do this on Linux, I can then run `$HOME/sandboxes/msb_tidb3_0_0/use` and I am connected to the database.
On MacOS, I get an error instead.

**To Reproduce**

1. run `dbdeployer deploy single tidb3.0.0 --client-from=5.7.25` 
2. run `$HOME/sandboxes/msb_tidb3_0_0/use`
3. See error "ERROR 1045 (28000): Access denied for user 'msandbox'@'localhost' (using password: YES)"

**Expected behavior**
The client connects

**Environment:**
 - OS: MacOs 10.14.2
 - dbdeployer version 1.18.0
 - tarball full name tidb-master-darwin-amd64.tar.gz
 - tarball origin https://download.pingcap.org/tidb-master-darwin-amd64.tar.gz

**Additional context**
The tarball was expanded manually, as the one provided was lacking the top directory.

I think there is a timing issue here, as the database server might not be ready when `load_grants` runs.

cc: @morgo

**Comment by morgo** [comment]
(Related) The Mac OS X packages have been fixed to create a top-level directory.  We can now recommend the following two paths for dbdeployer users:
- https://download.pingcap.org/tidb-master-linux-amd64.tar.gz
- https://download.pingcap.org/tidb-master-darwin-amd64.tar.gz

I will take a look on my Mac OS X machine and try to reproduce.

**Comment by datacharmer** [comment]
Is it possible to include the version in the tarball names?

**Comment by morgo** [comment]
These are trunk builds, so they are technically versionless.  They will advertise themselves when invoked as the **prior** version number of TiDB.

I managed to reproduce the bug.  It is **not a timing problem** (adding sleep 3 seconds did not fix it). 
 What has happened is that the script ran, but for some reason, the privileges didn't flush.

The workaround is:

```
FLUSH PRIVILEGES;
```

Syncronously flushing the privileges was only recently added to TiDB ([I added it here](https://github.com/pingcap/tidb/pull/8886/files)).  Previously it relied on a notification channel from pd, which doesn't exist in this deployment model.  I will have to check it out and see if there is something missing.

**Comment by morgo** [comment]
Is it possible that the `load_grants` script is double executing?  The log file complains about connections 2 and 3, and rolls back a connection due to an atuocommit error:
```
2019/02/17 09:31:22.752 session.go:1582: [info] [CRUCIAL OPERATION] con:1 schema_ver:15 set password for user <nil> (by root@localhost).
2019/02/17 09:31:22.753 session.go:1582: [info] [CRUCIAL OPERATION] con:1 schema_ver:15 create user {msandbox@127.% password = ***} (by root@localhost).
2019/02/17 09:31:22.754 session.go:1582: [info] [CRUCIAL OPERATION] con:1 schema_ver:15 grant all on *.* to msandbox@'127.%' (by root@localhost).
2019/02/17 09:31:22.755 session.go:1582: [info] [CRUCIAL OPERATION] con:1 schema_ver:15 create user {msandbox@localhost password = ***} (by root@localhost).
2019/02/17 09:31:22.755 session.go:1582: [info] [CRUCIAL OPERATION] con:1 schema_ver:15 grant all on *.* to msandbox@'localhost' (by root@localhost).
2019/02/17 09:31:22.756 session.go:1582: [info] [CRUCIAL OPERATION] con:1 schema_ver:15 create user {msandbox_rw@localhost password = ***} (by root@localhost).
2019/02/17 09:31:22.757 session.go:1582: [info] [CRUCIAL OPERATION] con:1 schema_ver:15 grant SELECT,INSERT,UPDATE,DELETE,CREATE,DROP,INDEX,ALTER,
    SHOW DATABASES,CREATE TEMPORARY TABLES,LOCK TABLES, EXECUTE
    on *.* to msandbox_rw@'localhost' (by root@localhost).
2019/02/17 09:31:22.760 session.go:1582: [info] [CRUCIAL OPERATION] con:1 schema_ver:15 create user {msandbox_rw@127.% password = ***} (by root@localhost).
2019/02/17 09:31:22.761 session.go:1582: [info] [CRUCIAL OPERATION] con:1 schema_ver:15 grant SELECT,INSERT,UPDATE,DELETE,CREATE,DROP,INDEX,ALTER,
    SHOW DATABASES,CREATE TEMPORARY TABLES,LOCK TABLES, EXECUTE
    on *.* to msandbox_rw@'127.%' (by root@localhost).
2019/02/17 09:31:22.764 session.go:1582: [info] [CRUCIAL OPERATION] con:1 schema_ver:15 create user {msandbox_ro@127.% password = ***} (by root@localhost).
2019/02/17 09:31:22.765 session.go:1582: [info] [CRUCIAL OPERATION] con:1 schema_ver:15 create user {msandbox_ro@localhost password = ***} (by root@localhost).
2019/02/17 09:31:22.765 session.go:1582: [info] [CRUCIAL OPERATION] con:1 schema_ver:15 create user {rsandbox@127.% password = ***} (by root@localhost).
2019/02/17 09:31:22.766 session.go:1582: [info] [CRUCIAL OPERATION] con:1 schema_ver:15 grant SELECT,EXECUTE on *.* to msandbox_ro@'127.%' (by root@localhost).
2019/02/17 09:31:22.767 session.go:1582: [info] [CRUCIAL OPERATION] con:1 schema_ver:15 grant SELECT,EXECUTE on *.* to msandbox_ro@'localhost' (by root@localhost).
2019/02/17 09:31:22.768 session.go:1582: [info] [CRUCIAL OPERATION] con:1 schema_ver:15 grant REPLICATION SLAVE on *.* to rsandbox@'127.%' (by root@localhost).
2019/02/17 09:31:22.768 session.go:1584: [info] [CRUCIAL OPERATION] con:1 schema_ver:15 cur_db:mysql create schema if not exists test (by root@localhost).
2019/02/17 09:31:22.768 server.go:316: [info] con:1 close connection
2019/02/17 09:31:22.799 server.go:314: [info] con:2 new connection 
2019/02/17 09:31:22.800 session.go:1582: [info] [CRUCIAL OPERATION] con:2 schema_ver:15 set password for user <nil> (by root@localhost).
2019/02/17 09:31:22.801 session.go:1582: [info] [CRUCIAL OPERATION] con:2 schema_ver:15 create user {msandbox@127.% password = ***} (by root@localhost).
2019/02/17 09:31:22.801 tidb.go:149: [info] RollbackTxn for ddl/autocommit error.
2019/02/17 09:31:22.801 session.go:872: [warning] con:2 schema_ver:15 session error:
Duplicate user
github.com/pingcap/tidb/executor.(*SimpleExec).executeCreateUser
        /Users/pingcap/gopkg/src/github.com/pingcap/tidb/executor/simple.go:169
github.com/pingcap/tidb/executor.(*SimpleExec).Next
        /Users/pingcap/gopkg/src/github.com/pingcap/tidb/executor/simple.go:68
github.com/pingcap/tidb/executor.(*ExecStmt).handleNoDelayExecutor
        /Users/pingcap/gopkg/src/github.com/pingcap/tidb/executor/adapter.go:296
github.com/pingcap/tidb/executor.(*ExecStmt).Exec
        /Users/pingcap/gopkg/src/github.com/pingcap/tidb/executor/adapter.go:242
github.com/pingcap/tidb/session.runStmt
        /Users/pingcap/gopkg/src/github.com/pingcap/tidb/session/tidb.go:195
github.com/pingcap/tidb/session.(*session).executeStatement
        /Users/pingcap/gopkg/src/github.com/pingcap/tidb/session/session.go:869
github.com/pingcap/tidb/session.(*session).execute
        /Users/pingcap/gopkg/src/github.com/pingcap/tidb/session/session.go:937
github.com/pingcap/tidb/session.(*session).Execute
        /Users/pingcap/gopkg/src/github.com/pingcap/tidb/session/session.go:890
github.com/pingcap/tidb/server.(*TiDBContext).Execute
        /Users/pingcap/gopkg/src/github.com/pingcap/tidb/server/driver_tidb.go:245
github.com/pingcap/tidb/server.(*clientConn).handleQuery
        /Users/pingcap/gopkg/src/github.com/pingcap/tidb/server/conn.go:1007
github.com/pingcap/tidb/server.(*clientConn).dispatch
        /Users/pingcap/gopkg/src/github.com/pingcap/tidb/server/conn.go:743
github.com/pingcap/tidb/server.(*clientConn).Run
        /Users/pingcap/gopkg/src/github.com/pingcap/tidb/server/conn.go:587
github.com/pingcap/tidb/server.(*Server).onConn
        /Users/pingcap/gopkg/src/github.com/pingcap/tidb/server/server.go:324
runtime.goexit
        /Users/pingcap/go/src/runtime/asm_amd64.s:1333
{
  "currDBName": "mysql",
  "id": 2,
  "status": 2,
  "strictMode": true,
  "user": {
    "Username": "root",
    "Hostname": "localhost",
    "CurrentUser": false,
    "AuthUsername": "root",
    "AuthHostname": "%"
  }
}
2019/02/17 09:31:22.801 conn.go:605: [warning] con:2 dispatch error:
id:2, addr: status:2, collation:utf8_general_ci, user:root
"create user msandbox@'127.%' identified by 'msandbox'"
Duplicate user
github.com/pingcap/tidb/executor.(*SimpleExec).executeCreateUser
        /Users/pingcap/gopkg/src/github.com/pingcap/tidb/executor/simple.go:169
github.com/pingcap/tidb/executor.(*SimpleExec).Next
        /Users/pingcap/gopkg/src/github.com/pingcap/tidb/executor/simple.go:68
github.com/pingcap/tidb/executor.(*ExecStmt).handleNoDelayExecutor
        /Users/pingcap/gopkg/src/github.com/pingcap/tidb/executor/adapter.go:296
github.com/pingcap/tidb/executor.(*ExecStmt).Exec
        /Users/pingcap/gopkg/src/github.com/pingcap/tidb/executor/adapter.go:242
github.com/pingcap/tidb/session.runStmt
        /Users/pingcap/gopkg/src/github.com/pingcap/tidb/session/tidb.go:195
github.com/pingcap/tidb/session.(*session).executeStatement
        /Users/pingcap/gopkg/src/github.com/pingcap/tidb/session/session.go:869
github.com/pingcap/tidb/session.(*session).execute
        /Users/pingcap/gopkg/src/github.com/pingcap/tidb/session/session.go:937
github.com/pingcap/tidb/session.(*session).Execute
        /Users/pingcap/gopkg/src/github.com/pingcap/tidb/session/session.go:890
github.com/pingcap/tidb/server.(*TiDBContext).Execute
        /Users/pingcap/gopkg/src/github.com/pingcap/tidb/server/driver_tidb.go:245
github.com/pingcap/tidb/server.(*clientConn).handleQuery
        /Users/pingcap/gopkg/src/github.com/pingcap/tidb/server/conn.go:1007
github.com/pingcap/tidb/server.(*clientConn).dispatch
        /Users/pingcap/gopkg/src/github.com/pingcap/tidb/server/conn.go:743
github.com/pingcap/tidb/server.(*clientConn).Run
        /Users/pingcap/gopkg/src/github.com/pingcap/tidb/server/conn.go:587
github.com/pingcap/tidb/server.(*Server).onConn
        /Users/pingcap/gopkg/src/github.com/pingcap/tidb/server/server.go:324
runtime.goexit
        /Users/pingcap/go/src/runtime/asm_amd64.s:1333
2019/02/17 09:31:22.801 server.go:316: [info] con:2 close connection
2019/02/17 09:31:22.824 server.go:314: [info] con:3 new connection 
2019/02/17 09:31:22.824 server.go:314: [info] con:3 new connection 
2019/02/17 09:31:22.825 session.go:1582: [info] [CRUCIAL OPERATION] con:3 schema_ver:15 set password for user <nil> (by root@localhost).
2019/02/17 09:31:22.825 session.go:1582: [info] [CRUCIAL OPERATION] con:3 schema_ver:15 create user {msandbox@127.% password = ***} (by root@localhost).
2019/02/17 09:31:22.825 tidb.go:149: [info] RollbackTxn for ddl/autocommit error.
2019/02/17 09:31:22.826 session.go:872: [warning] con:3 schema_ver:15 session error:
Duplicate user
github.com/pingcap/tidb/executor.(*SimpleExec).executeCreateUser
        /Users/pingcap/gopkg/src/github.com/pingcap/tidb/executor/simple.go:169
github.com/pingcap/tidb/executor.(*SimpleExec).Next
        /Users/pingcap/gopkg/src/github.com/pingcap/tidb/executor/simple.go:68
github.com/pingcap/tidb/executor.(*ExecStmt).handleNoDelayExecutor
        /Users/pingcap/gopkg/src/github.com/pingcap/tidb/executor/adapter.go:296
github.com/pingcap/tidb/executor.(*ExecStmt).Exec
        /Users/pingcap/gopkg/src/github.com/pingcap/tidb/executor/adapter.go:242
github.com/pingcap/tidb/session.runStmt
        /Users/pingcap/gopkg/src/github.com/pingcap/tidb/session/tidb.go:195
github.com/pingcap/tidb/session.(*session).executeStatement
        /Users/pingcap/gopkg/src/github.com/pingcap/tidb/session/session.go:869
github.com/pingcap/tidb/session.(*session).execute
        /Users/pingcap/gopkg/src/github.com/pingcap/tidb/session/session.go:937
github.com/pingcap/tidb/session.(*session).Execute
        /Users/pingcap/gopkg/src/github.com/pingcap/tidb/session/session.go:890
github.com/pingcap/tidb/server.(*TiDBContext).Execute
        /Users/pingcap/gopkg/src/github.com/pingcap/tidb/server/driver_tidb.go:245
github.com/pingcap/tidb/server.(*clientConn).handleQuery
        /Users/pingcap/gopkg/src/github.com/pingcap/tidb/server/conn.go:1007
github.com/pingcap/tidb/server.(*clientConn).dispatch
        /Users/pingcap/gopkg/src/github.com/pingcap/tidb/server/conn.go:743
github.com/pingcap/tidb/server.(*clientConn).Run
        /Users/pingcap/gopkg/src/github.com/pingcap/tidb/server/conn.go:587
github.com/pingcap/tidb/server.(*Server).onConn
        /Users/pingcap/gopkg/src/github.com/pingcap/tidb/server/server.go:324
runtime.goexit
        /Users/pingcap/go/src/runtime/asm_amd64.s:1333
{
  "currDBName": "mysql",
  "id": 3,
  "status": 2,
  "strictMode": true,
  "user": {
    "Username": "root",
    "Hostname": "localhost",
    "CurrentUser": false,
    "AuthUsername": "root",
    "AuthHostname": "%"
  }
}
2019/02/17 09:31:22.826 conn.go:605: [warning] con:3 dispatch error:
id:3, addr: status:2, collation:utf8_general_ci, user:root
"create user msandbox@'127.%' identified by 'msandbox'"
Duplicate user
github.com/pingcap/tidb/executor.(*SimpleExec).executeCreateUser
        /Users/pingcap/gopkg/src/github.com/pingcap/tidb/executor/simple.go:169
github.com/pingcap/tidb/executor.(*SimpleExec).Next
        /Users/pingcap/gopkg/src/github.com/pingcap/tidb/executor/simple.go:68
github.com/pingcap/tidb/executor.(*ExecStmt).handleNoDelayExecutor
        /Users/pingcap/gopkg/src/github.com/pingcap/tidb/executor/adapter.go:296
github.com/pingcap/tidb/executor.(*ExecStmt).Exec
        /Users/pingcap/gopkg/src/github.com/pingcap/tidb/executor/adapter.go:242
github.com/pingcap/tidb/session.runStmt
        /Users/pingcap/gopkg/src/github.com/pingcap/tidb/session/tidb.go:195
github.com/pingcap/tidb/session.(*session).executeStatement
        /Users/pingcap/gopkg/src/github.com/pingcap/tidb/session/session.go:869
github.com/pingcap/tidb/session.(*session).execute
        /Users/pingcap/gopkg/src/github.com/pingcap/tidb/session/session.go:937
github.com/pingcap/tidb/session.(*session).Execute
        /Users/pingcap/gopkg/src/github.com/pingcap/tidb/session/session.go:890
github.com/pingcap/tidb/server.(*TiDBContext).Execute
        /Users/pingcap/gopkg/src/github.com/pingcap/tidb/server/driver_tidb.go:245
github.com/pingcap/tidb/server.(*clientConn).handleQuery
        /Users/pingcap/gopkg/src/github.com/pingcap/tidb/server/conn.go:1007
github.com/pingcap/tidb/server.(*clientConn).dispatch
        /Users/pingcap/gopkg/src/github.com/pingcap/tidb/server/conn.go:743
github.com/pingcap/tidb/server.(*clientConn).Run
        /Users/pingcap/gopkg/src/github.com/pingcap/tidb/server/conn.go:587
github.com/pingcap/tidb/server.(*Server).onConn
        /Users/pingcap/gopkg/src/github.com/pingcap/tidb/server/server.go:324
runtime.goexit
        /Users/pingcap/go/src/runtime/asm_amd64.s:1333
2019/02/17 09:31:22.826 server.go:316: [info] con:3 close connection
```

**Comment by datacharmer** [comment]
The script is actually called three times.
one with argument `pre_grants.sql`, one without argument (defaults to `grants.mysql`), and one with argument `post_grants.sql`

If the script does not exist, the original `load_grants` exit without executing anything.
I can fix the actual TiDB script to reproduce this behavior.

**Comment by datacharmer** [comment]
I wonder why it doesn't fail on Linux, though

**Comment by morgo** [comment]
It is by nature a race condition.  But I can't explain why it reliably triggers on Mac OS and not Linux.  What do you think about the workaround of just adding `FLUSH PRIVILEGES` to the end of the script?

**Comment by datacharmer** [comment]
I can do that, in addition to adding the proper handling of `pre-` and `post-` arguments

**Comment by datacharmer** [comment]
I have added `FLUSH PRIVILEGES`, and I realised that the original load_grants will work just fine, without double load.
I moved the unwanted scripts removal to a dedicated place (a new script named "after_start"),  and now the installation works on both MacOS and Linux.
I had to leave the small sleep in place, though. It fails randomly without. This is matter for further investigation.
Will commit soon.
